jgrapht-core/src/main/java/org/jgrapht/generate/netgen/BipartiteMatchingProblem.java

```diff
@@ -0,0 +1,159 @@
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the
+ * GNU Lesser General Public License v2.1 or later
+ * which is available at
+ * http://www.gnu.org/licenses/old-licenses/lgpl-2.1-standalone.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.generate.netgen;
+
+import org.jgrapht.Graph;
+
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * This class represents a bipartite matching problem. The problem can be weighted or
+ * unweighted depending on the {@link BipartiteMatchingProblem#isWeighted()}.
+ * <p>
+ * The minimum weight (minimum cost) perfect bipartite matching problem is defined as follows:
+ * \[ \begin{align} \mbox{minimize}~&amp; \sum_{e \in E}c_e\cdot x_e &amp;\\
+ * \mbox{s.t. }&amp;\sum_{e\in \delta(v)} x_e = 1 &amp; \forall v\in V\\
+ * &amp;x_e \in \{0,1\} &amp; \forall e\in E \end{align} \]
+ * Here $\delta(v)$ denotes the set of edges incident to the vertex $v$.
+ * The parameters $c_{e}$ define a cost of adding the edge $e$ to the matching.
+ * If the problem is unweighted, the values $c_e$ are equal to 1 in the problem formulation.
+ * <p>
+ * This class can define bipartite matching problems without the requirement that every
+ * edge must be matched, i.e. non-perfect matching problems. These problems are called
+ * maximum cardinality bipartite matching problems. The goal of the maximum cardinality
+ * matching problem is to find a matching with maximum number of edges. If the cost
+ * function is used in this setup, the goal is to find the cheapest matching among all
+ * matchings of maximum cardinality.
+ *
+ * @param <V> the graph vertex types
+ * @param <E> the graph edge type
+ * @author Timofey Chudakov
+ * @see org.jgrapht.alg.interfaces.MatchingAlgorithm
+ */
+public interface BipartiteMatchingProblem<V, E> {
+
+    /**
+     * Returns the graph, which defines the problem
+     *
+     * @return the graph, which defines the problem
+     */
+    Graph<V, E> getGraph();
+
+    /**
+     * Returns one of the 2 partitions of the graph (no 2 vertices in this set share an edge)
+     *
+     * @return one of the 2 partitions of the graph
+     */
+    Set<V> getPartition1();
+
+    /**
+     * Returns one of the 2 partitions of the graph (no 2 vertices in this set share an edge)
+     *
+     * @return one of the 2 partitions of the graph
+     */
+    Set<V> getPartition2();
+
+    /**
+     * Returns a cost function of this problem. This function must be defined for all edges
+     * of the graph. In the case the problem is unweighted, the function must return any
+     * constant value for all edges.
+     *
+     * @return a cost function of this problem
+     */
+    Function<E, Integer> getCosts();
```

Comment:
It is always an integer?

Reply:
It was a default solution. In fact, this is a point to discuss because there're a number of alternative solutions. 
* parameterize the problem with the number data type it uses (`<..., C extends Number>`).
* make every function to return `int`
* make every function to return `double`
* make cost functions return `double` and capacity function return `int` (as in `MinimumCostFlowProblem`).

Maximum flow and matching algorithms currently work with double graph edge weights, so solution 3 seems to be the most convenient except for that it isn't aligned with the `MinimumCostFlowProblem`. So, should we go with number 3?

Reply:
So this generator always returns integer weights for the bipartite problems?

I guess it is fine this way, the user will have to do the conversion before calling for example the maximum weight matching algorithm.